### DISCUSSION TITLE
Setting PPQN Based Midi Timing

### DISCUSSION THREAD

**Discussion by DanAudio** [DBraun/DawDreamer#161] [Q&A]
I am performing brute forcing with DawDreamer, and it is working well. However, I have an issue with handling MIDI timing, which seems synchronized to the audio buffer. I can't synchronize to DAWs which use PPQN based timing.

Is there a way to synchronize MIDI timing to PPQN as with automation instead of using the audio block?

**Comment by DBraun** [comment]
I think the features are there.

https://github.com/DBraun/DawDreamer/wiki/Render-Engine-and-BPM -> You can set a BPM as an automation.
```python
engine.set_bpm(bpm_automation, ppqn=PPQN)
```
That would be like drawing an automation curve for the BPM in a DAW. Every PPQN (such as 960) samples in the automation array correspond to a beat of music (0.5 seconds at 120 bpm)

Second, you can tie individual MIDI notes or entire MIDI files to beats: https://github.com/DBraun/DawDreamer/wiki/Plugin-Processor
```python
synth.load_midi(MIDI_PATH, beats=True)
synth.add_midi_note(67, 127, 1, .5, beats=True)
```

If something's not clear let me know, or share a script of what you're working on.

**Reply by DanAudio** [reply]
Thank you for the swift response DBraun.

With large buffer sizes (for example 4096) MIDI sounds choppy even after running the script (with the BPM changed to the desired BPM). MIDI is set to beats. Are the MIDI note on/offs tied to the automation curve and not the buffer?

**Reply by DBraun** [reply]
Does it sound better with a small buffer size? Can you try 128?

The MIDI note on/offs are tied to the automation curve if the `beats` keyword is True when you call either `load_midi` or `add_midi_note`.

**Reply by DanAudio** [reply]
Yes, it sounds better with small buffer, the beats keyword is set to True on load_midi.

Is it possible the MIDI note on/offs are not delivered to the buffer before processing?

**Reply by DBraun** [reply]
Yea it could be a bug. Are you using `load_midi` or `add_midi_note`?

**Reply by DanAudio** [reply]
I have only used load_midi so far.
